lib/basedriver/image-helpers.js

```diff
@@ -34,29 +49,39 @@ async function createBase64FromImage (image) {
   });
 }
 
+/**
+ * Crop the image by given rectangle
+ *
+ * @param {PNG} image The image to crop
+ * @param {{left: number, top: number, width: number, height: number}} rect The selected region of image
+ * @return {PNG} Cropped image
+ */
 function cropImage (image, rect) {
   const imageRect = {width: image.width, height: image.height};
   const interRect = getRectIntersection(rect, imageRect);
   if (isEmptyRect(interRect)) {
     throw new Error('rect is outside the image bounds!');
   }
 
-  const yEnd = Math.min(interRect.top + interRect.height, imageRect.height);
-  const xEnd = Math.min(interRect.left + interRect.width, imageRect.width);
+  const firstVerticalPixel = interRect.top;
+  const lastVerticalPixel = interRect.top + interRect.height;
+
+  const firstHorizontalPixel = interRect.left;
+  const lastHorizontalPixel = interRect.left + interRect.width;
 
-  let croppedArray = [];
-  for (let y = interRect.top; y < yEnd; y++) {
-    for (let x = interRect.left; x < xEnd; x++) {
-      const idx = (imageRect.width * y + x) << 2;
-      for (let i = 0; i < 4; i++) {
-        croppedArray.push(image.data[idx + i]);
+  const croppedArray = [];
+  for (let y = firstVerticalPixel; y < lastVerticalPixel; y++) { // loop over vertical pixels
+    for (let x = firstHorizontalPixel; x < lastHorizontalPixel; x++) { // loop over horizontal pixels
+      const idx = (imageRect.width * y + x) << 2; // first byte location of current pixel's block
+      for (let i = 0; i < 4; i++) { // each pixel's block consists of 4 bytes
```

Comment:
proper variable names will make these comments unnecessary, for example

`for (let i = 0; i < 4; i++)` -> `for (let byteIdx = 0; byteIdx < BYTES_IN_BLOCK; ++byteIdx)`